ouroboros-network/src/Ouroboros/Network/AnchoredFragment.hs

```diff
@@ -0,0 +1,554 @@
+{-# LANGUAGE PatternSynonyms     #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies        #-}
+{-# LANGUAGE ViewPatterns        #-}
+module Ouroboros.Network.AnchoredFragment (
+  -- * AnchoredFragment type and fundamental operations
+  AnchoredFragment(Empty, (:>), (:<)),
+  anchorPoint,
+  unanchorFragment,
+  mkAnchoredFragment,
+  valid,
+  validExtension,
+
+  -- ** Block re-exports
+  HasHeader(..),
+  Point(..),
+  castPoint,
+  blockPoint,
+
+  -- * AnchoredFragment construction and inspection
+  -- ** Head inspection
+  headPoint,
+  headSlot,
+  headHash,
+  headBlockNo,
+
+  -- ** Basic operations
+  head,
+  last,
+  lastPoint,
+  lastSlot,
+  toNewestFirst,
+  toOldestFirst,
+  fromNewestFirst,
+  fromOldestFirst,
+  dropNewest,
+  takeOldest,
+  dropWhileNewest,
+  length,
+  null,
+
+  -- ** Update type and operations
+  ChainUpdate(..),
+  addBlock,
+  rollback,
+  applyChainUpdate,
+  applyChainUpdates,
+
+  -- * Special operations
+  pointOnFragment,
+  withinFragmentBounds,
+  findFirstPoint,
+  successorBlock,
+  selectPoints,
+  isPrefixOf,
+  join,
+  intersect,
+  intersectionPoint,
+
+  -- * Conversion to/from Chain
+  fromChain,
+  toChain,
+  anchorNewest,
+
+  -- * Helper functions
+  prettyPrint
+  ) where
+
+import           Prelude hiding (head, last, length, null)
+
+import           Control.Exception (assert)
+import           Data.List (find)
+import           Data.Word (Word64)
+
+import           Ouroboros.Network.Block
+import           Ouroboros.Network.Chain (Chain)
+import qualified Ouroboros.Network.Chain as Chain
+import           Ouroboros.Network.ChainFragment (ChainFragment)
+import qualified Ouroboros.Network.ChainFragment as CF
+
+
+-- | An 'AnchoredFragment' is a 'ChainFragment' that is anchored to some start
+-- 'Point': we know the 'Point' right before the first, leftmost block in the
+-- fragment. The anchor point can be thought of as a left exclusive bound.
+--
+-- For example, the following fragment is anchored at @a@ and contains @b1@,
+-- @b2@, and @b3@, which is the head of the fragment.
+--
+-- > a ] b1 >: b2 >: b3
+--
+-- The fact that it is an /exclusive/ bound is particularly convenient when
+-- dealing with Genesis, which is the start of the chain, but not an actual
+-- block, but there /is/ a 'Point' that refers to Genesis
```

Comment:
slightly awkward sentence construction here... "which is.. but not an... but there is..".